File: go/builtins/funcFromJSON.go
package builtins

import (
	"encoding/json"

	"github.com/2manyvcos/jpl/go/jpl"
	"github.com/2manyvcos/jpl/go/library"
)

var funcFromJSON jpl.JPLFunc = func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	value, err := library.UnwrapValue(input)
	if err != nil {
		return nil, err
	}
	t, err := library.Type(value)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_STRING:
		var result any
		err := json.Unmarshal([]byte(value.(string)), &result)
		if err != nil {
			return nil, library.ThrowAny(library.NewRuntimeError(err.Error()))
		}
		return next.Pipe(result)

	default:
	}

	return nil, library.ThrowAny(library.NewTypeError("%s (%*<100v) cannot be parsed as JSON", string(t), value))
}

File: js/lib/builtins/funcFromJSON.ts
import {
  JPLRuntimeError,
  JPLTypeError,
  type JPLRuntimeSignal,
} from '@/library';
import type JPLRuntime from '@/runtime';

export default function builtin(
  runtime: JPLRuntime,
  _signal: JPLRuntimeSignal,
  next: (output: unknown) => Promise<unknown[]> | unknown[],
  input: unknown,
): Promise<unknown[]> | unknown[] {
  const value = runtime.unwrapValue(input);
  const t = runtime.type(value);

  switch (t) {
    case 'string': {
      let result;
      try {
        result = JSON.parse(value as string);
      } catch (err) {
        throw new JPLRuntimeError((err as Error).message);
      }
      return next(result);
    }

    default:
  }

  throw new JPLTypeError('%s (%*<100v) cannot be parsed as JSON', t, value);
}
